src/wasm-stack.h

```diff
@@ -197,10 +197,21 @@ template<typename SubType> void BinaryenIRWriter<SubType>::write() {
   emitFunctionEnd();
 }
 
-// emits a node, but if it is a block with no name, emit a list of its contents
+// Emits a node in a position that can contain a list of contents, like an if
+// arm. This will emit the node, but if it is a block with no name, just emit
+// its contents. This is ok to do because a list of contents is ok in the wasm
+// binary format in such positions anyhow. When we read such code in Binaryen
+// we will end up creating a block for it (note that while doing so we create a
+// block without a name, since nothing branches to it, which makes it easy to
+// handle in optimization passes and when writing the binary out again).
 template<typename SubType>
 void BinaryenIRWriter<SubType>::visitPossibleBlockContents(Expression* curr) {
   auto* block = curr->dynCast<Block>();
+  // Even if the block has a name, check if the name is necessary (if it has no
+  // uses, it is equivalent to not having one). This is potentially quadratic,
+  // but it is extremely rare to have recursion on this function, since it is
+  // limited by the number of non-block control flow structures (the places that
+  // call here).
```

Comment:
> This is potentially quadratic,

Isn't this _always_ quadratic, because we traverse all children?

> it is extremely rare to have recursion on this function, since it is limited by the number of non-block control flow structures (the places that call here).

How do we have a recursion here? Can `BranchSeeker::has` call this `BinaryenIRWriter::VisitPossibleBlockContents` back or something?

Reply:
Hmm, good points, my text was not accurate. Yes, it's always quadratic, just on smaller numbers. By "recursion" I really meant larger numbers (where quadratic time gets bad). I improved the comment now.